File: prisma/refs/py/generate-parentheses.py
def generate_parenthesis(n):
    res = []

    def bt(cur, open_, close):
        if len(cur) == 2 * n:
            res.append(cur)
            return
        if open_ < n:
            bt(cur + "(", open_ + 1, close)
        if close < open_:
            bt(cur + ")", open_, close + 1)

    bt("", 0, 0)
    return res

File: prisma/refs/generate-parentheses.ts
export function generateParenthesis(n: number): string[] {
  const res: string[] = [];
  const bt = (cur: string, open: number, close: number) => {
    if (cur.length === 2 * n) {
      res.push(cur);
      return;
    }
    if (open < n) bt(cur + "(", open + 1, close);
    if (close < open) bt(cur + ")", open, close + 1);
  };
  bt("", 0, 0);
  return res;
}
